### ISSUE TITLE
Support local imports (use statements in blocks)

### ISSUE THREAD

**Issue by suhr** [rust-analyzer/rust-analyzer#1165]
Inside a function, I often write something like this:

```rust
use Request as R;

match ... {
    R::Info => {...},
    ...
}
```

But RA seems to have a problem with such imports:

```
Failed to exactly resolve the symbol. This is probably because rust_analyzer does not yet support traits.
These items were found instead:

Request
```

**Comment by edwin0cheng** [comment]
Yeah, Seem like current [`ra_hir::expr::Body`](https://github.com/rust-analyzer/rust-analyzer/blob/d55f1136d6444b1f50b9092c36a976d0e1c26202/crates/ra_hir/src/expr.rs#L29) do not store local items information :
```
pub struct Body {
    /// The def of the item this body belongs to
    owner: DefWithBody,
    exprs: Arena<ExprId, Expr>,
    pats: Arena<PatId, Pat>,
    /// The patterns for the function's parameters. While the parameter types are
    /// part of the function signature, the patterns are not (they don't change
    /// the external type of the function).
    ///
    /// If this `Body` is for the body of a constant, this will just be
    /// empty.
    params: Vec<PatId>,
    /// The `ExprId` of the actual body expression.
    body_expr: ExprId,
}
```

An here is a simplifed test case for this bug:

```
enum Request {
    Info
}

fn main() {
    use Request as R;    
    let _r = R::Info;
}
```
```
hover on _r = {unknown}
```

**Comment by matklad** [comment]
cc @pnkfelix, supporting local items (\w imports) and laziness is going to be a major design constarint on the name resolution library.

The current code in rust-analyzer just completely ignores local items and builds a Resolver instance for top-level scope. 

We can ask the Resolver to process item bodies at the same time as it processes modules, but that would kill laziness. 

Instead, I hope that for each function we can create a separate Resolver instance on-demand which references the parent resolver. That is, each function gets its own independent resolver, but they all share the same module-level scope

**Comment by ice1000** [comment]
Is this planned to be implemented (yet)?

**Comment by matklad** [comment]
This is not planned for the nearest future

**Comment by oblitum** [comment]
Just avoided creating another duplicate of this... Give the amount of refs and duplicates this looks at top priority.

**Comment by Dushistov** [comment]
`use Enum::*` inside function also covered by this issue?

**Comment by ice1000** [comment]
> `use Enum::*` inside function also covered by this issue?
> Or it is covered by #4749 ?

I guess the answer is "both"

**Comment by flodiebold** [comment]
Any `use` inside a function does not work at all, because of this issue. #4749 is a specific issue with an assist.

**Comment by ice1000** [comment]
No offense, but it's quite disappointing seeing this issue as not updated nearly a year after the last time I use ra..... It may affect the core design, so imho doing this as early as possible can minimize the amount of code we're gonna change when the refactoring is performed.

**Comment by kiljacken** [comment]
No offense, but it's quite disappointing to see people leaving negatively loaded comments about their pet feature not being the prime focus of the project.

Sorry if that came across as condescending, but I needed to get it off my chest and make a point about how to word a message. Keep things positive and you're way more likely to get a positive reaction.

We'd be more than happy to accept any PR's that help the issue along, but so far focus has just been elsewhere because this issue is very workaround-able, and especially because it need a fair bit of design thought and planning, while there is still low-hanging fruit that improves the experience elsewhere.

**Comment by lnicola** [comment]
I think this and the related #1559 are the most duplicated issues on this tracker, so it's clear they are important. On the other hand, they're not trivial to implement and nobody has put in the effort or the funding for it. They will be implemented, but other foundational work like sharing the parser between the compiler and `rust-analyzer` is more important.

**Comment by flodiebold** [comment]
Matklad actually recently wrote an issue for the required refactorings: #5922

**Comment by ice1000** [comment]
> Matklad actually recently wrote an issue for the required refactorings: #5922

Oh, that's cool!

**Comment by MikailBag** [comment]
If RA desugars each function with local imports into a module with `use` and function within it, will it work?

**Comment by lnicola** [comment]
@MikailBag you can have use statements at scope level.

**Comment by flodiebold** [comment]
It's not that simple, the anonymous modules created by block-level items work differently from named modules (in that everything from the containing named module stays in scope).

**Comment by flodiebold** [comment]
Also, the main complication here is not actually implementing the logic, but doing it in a way that doesn't require us to parse all function bodies all the time.

**Comment by ice1000** [comment]
> If RA desugars each function with local imports into a module with `use` and function within it, will it work?

This will produce significant large amount of modules, which is no good.

**Comment by matklad** [comment]
Yeah, the #5922 is still how I think we should proceed here. Basically, we first should refactor existing code to be more modular, and than we should add support for local modules. 

PRs welcome, but be warned that this is a large task. I'd estimate it as 1-2 weeks full-time for someone who is already familiar with the relevant infra.

At the same time, I wouldn't worry to much about "core design" here. I think both of these are true:

* the whole `DefCollector` thing needs to be rewritten
* this won't affect code outside of the name resolution, as the interface is isolated enough. 

Also, +1 to [this comment](https://github.com/rust-analyzer/rust-analyzer/issues/1165#issuecomment-696560892) regarding the language :-)

**Comment by djc** [comment]
At the risk of dragging this on more, I just want to make the point that from the perspective of this particular user, false positive compilation errors would have the most positive impact on my experience using RA -- substantially more than refactoring tools, search & replace tools, further performance improvements or more code sharing. Of course this disregards the experience of RA contributors and wider ecosystem benefits from things like code sharing, but I feel like it should count for something.

I don't know if/how RA prioritizes work currently -- making that more transparent might also help in the sense of making it more concrete what trade-offs are affecting issues like this. Also, this is the kind of thing where I wonder if a focused fundraising effort could help -- I for one would gladly contribute some cash to help prioritize this, but I don't think my one-man business can currently sustain paying for the entirety of 1-2 person-weeks of work.

**Comment by ice1000** [comment]
For wording, my initial intention was to bring some "ping"-like comments so there can be some discussion. This issue is so important but is only widely mentioned, and is on page 3 when sorting issues with "most discussed". I apologize for the moronic words though, I have made some attempts on trying to make this work (like introducing block-modules) before, but they turns out to be not good design, so I'm aware of the difficulty of the task.

Thanks for everyone who have made RA so great!

**Comment by emilk** [comment]
I'd just like to chime in with some motivation on why I think supporting local `use` statements is an important issue:

Increasingly I only put a `use` statement at the top of a file if the imported symbol will be used throughout the file. Instead I prefer my `use`-statements inside my functions. I find that this improves the readability of my code as it makes it more obvious where a certain symbol came from (without the reader having to scroll to the top of the file). But maybe even more importantly it improves refactorability (yes, it is a word). Earlier when moving code between files I needed to spend a lot of time figuring out which `use` statements (and before that `#include`s) to copy and/or move to the other file. When the `use` statements are part of the function body, that work is almost completely gone.

In short: just like I try to limit the scope of my variables, I try to limit the scope of my `use` statements.

**Comment by jonas-schievink** [comment]
This is now implemented and was fixed by https://github.com/rust-analyzer/rust-analyzer/pull/7614 / https://github.com/rust-analyzer/rust-analyzer/pull/7627.

There are still some known bugs and missing parts, like local `impl`s (which should at least work for types that are also local), and some remaining cleanups to be done, but the bulk of the work is now done, so closing.

**Comment by umanwizard** [comment]
It's so beautiful...

![image](https://user-images.githubusercontent.com/1893159/107567953-59371700-6bb4-11eb-90c4-ad9cb7541801.png)

Thanks @jonas-schievink ! this has been my most-desired missing feature in rust-analyzer for such a long time!

**Comment by djc** [comment]
@jonas-schievink woo! Thanks so much for picking this up.
